src/Router/PreloadLink.tsx

```diff
@@ -0,0 +1,146 @@
+      relayEnvironment: PropTypes.object.isRequired,
+      reactionRouter: PropTypes.shape({
+        routeConfig: PropTypes.array.isRequired,
+        resolver: PropTypes.object.isRequired,
+      }).isRequired,
+    }
+
+    static defaultProps = {
+      immediate: false,
+    }
+
+    state = {
+      isLoading: false,
+    }
+
+    componentDidMount() {
+      if (this.props.immediate) {
+        this.fetchData()
+      }
+    }
+
+    getRouteQuery() {
+      const {
+        reactionRouter: { resolver },
+        router,
+        to,
+      } = this.props
+
+      const { getRouteMatches, getRouteValues } = Found.ResolverUtils
+      const location = router.createLocation(to)
+      const match = router.matcher.match(location)
+      const routes = router.matcher.getRoutes(match)
+      const augmentedMatch = { ...match, routes }
+      const routeMatches = getRouteMatches(augmentedMatch)
+
+      const query = getRouteValues(
+        routeMatches,
+        route => route.getQuery,
+        route => route.query
+      ).find(q => !isUndefined(q))
+
+      const cacheConfig = getRouteValues(
+        routeMatches,
+        route => route.getCacheConfig,
+        route => route.cacheConfig
+      ).find(caches => !isUndefined(caches))
+
+      const routeVariables = resolver
+        .getRouteVariables(match, routeMatches)
+        .find(variables => !isUndefined(variables) && !isEmpty(variables))
+
+      return {
+        query,
+        cacheConfig,
+        routeVariables,
+      }
+    }
+
+    fetchData() {
+      const { query, cacheConfig, routeVariables } = this.getRouteQuery()
+      const { relayEnvironment } = this.props
+
+      // Prime the store cache
+      return new Promise(async (resolve, reject) => {
+        const missingEnvironmentOrQuery = !(relayEnvironment && query)
+        if (missingEnvironmentOrQuery) {
+          return resolve()
+        }
+        try {
+          await fetchQuery(relayEnvironment, query, routeVariables, cacheConfig)
+          resolve()
+        } catch (error) {
+          console.error("[isomorphic-found-relay] PreloadLink.js Error:", error)
+        }
+      })
+    }
+
+    handleClick = event => {
+      event.preventDefault()
+
+      this.setState({
+        isLoading: true,
+      })
+
+      this.fetchData().then(() => {
+        const { router, replace, to } = this.props
+
+        this.setState({
+          isLoading: false,
+        })
+
+        if (replace) {
+          router.replace(to)
+        } else {
+          router.push(to)
+        }
+      })
+    }
+
+    render() {
+      const _props = omit(
+        [
+          "currentUser",
+          "immediate",
+          "relayEnvironment",
+          "reactionRouter",
+          "routeConfig",
+        ],
+        this.props
+      )
+
+      return (
+        <Found.Link onClick={this.handleClick} {..._props}>
+          {this.props.children}
+          {this.state.isLoading ? " [loading...]" : null}
```

Comment:
This will eventually call back to `props.onLoading`, and apparently `react-relay-modern-network-layer` has `onProgress` functionality. This will improve UX.